### PR TITLE
add persistent state for integrations

### PR DESCRIPTION
**Pull request by zandrewitte**
### Step 3: Did you add / update tests for your changes in the right area?
- [ ] Unit/Component test
- [ ] Integration test	

### Step 4: I'm confident that everything is properly tested:
I got a PO / QA Approval by:
- [ ] Name

### Step 5: Can we ship this feature to production?
- [ ] Yes, I'm proud of my work. Ship it! :ship:

### CODE REVIEW

File: stackstate_checks_base/stackstate_checks/base/utils/persistent_state.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# (C) StackState 2020
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+
+import os
+import json
+from schematics.models import Model
```
Comment (hruhek) [inline stackstate_checks_base/stackstate_checks/base/utils/persistent_state.py:7]: this line should look like this `from schematics import Model` as `Model` is imported in `schematics.__init__py`

File: stackstate_checks_dev/stackstate_checks/dev/plugin/plugin.py [hunk 2]
```diff
@@ -187,3 +191,28 @@ def run_check(config=None, **kwargs):
...
+            self.persistent_state = StateManager(logger)
+
+        def assert_state(self, instance, state, state_schema=None):
+            assert self.persistent_state.get_state(instance, state_schema) is None
+            self.persistent_state.set_state(instance, state)
+            assert self.persistent_state.get_state(instance, state_schema) == state
+            self.persistent_state.flush(instance)
+            self.persistent_state.read_state(instance)
+            assert self.persistent_state.get_state(instance, state_schema) == state
+            self.persistent_state.clear(instance)
```
Comment (TizianoPerrucci) [inline stackstate_checks_dev/stackstate_checks/dev/plugin/plugin.py:212]: split this part so we can reuse it

File: stackstate_checks_base/tests/test_utils.py [hunk 3]
```diff
@@ -119,3 +127,91 @@ def test_round_modify_sig_digits(self):
...
+        s = TestStorageSchema({'offset': 10})
+        instance = StateDescriptor("on.disk.state.schema", ".")
+        rs = state.assert_state(instance, s, TestStorageSchema)
+        assert rs.offset == s.offset
+
+    def test_rollback_state(self, state):
+        s = TestStorageSchema({'offset': 10})
+        instance = StateDescriptor("rollback.state.schema", ".")
+
+        assert state.persistent_state.get_state(instance, TestStorageSchema) is None
```
Comment (TizianoPerrucci) [inline stackstate_checks_base/tests/test_utils.py:200]: reuse the fixture here

**Review by TizianoPerrucci — CHANGES_REQUESTED** [review]
We also talked about the possibility to have a `AgentStatefullCheck` class that can be extended by check that need state. It can provide awareness into the base class that the check uses state and can do proper check and state creation before the `check.run`.
Right now the base class does not have any clue if the check wanna use state and we defer the state write till the end after the `check.run` so we will know if there are any problems in writing files only at the end.

File: stackstate_checks_base/stackstate_checks/base/utils/persistent_state.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            self.flush(instance)
+
+    def flush(self, instance):
+        """
+        flush writes the state data for this instance to disk
+        `instance` the persistence instance for which the state is flushed to disk.
+        """
+        if instance.instance_key in self.state:
+            # check if folder and file exists before writing
+            if not os.path.exists(os.path.dirname(instance.file_location)):
```
Comment (TizianoPerrucci) [inline stackstate_checks_base/stackstate_checks/base/utils/persistent_state.py:137]: combine in one `try` with the `write` below

File: stackstate_checks_dev/stackstate_checks/dev/plugin/plugin.py [hunk 5]
```diff
@@ -187,3 +192,67 @@ def run_check(config=None, **kwargs):
...
+@pytest.fixture
+def state():
+    logger = logging.getLogger(__name__)
+
+    class PersistentStateFixture:
+
+        def __init__(self):
+            self.persistent_state = StateManager(logger)
+
+        def assert_state_check(self, check, pre_run_state, post_run_state, state_schema=None):
```
Comment (TizianoPerrucci) [inline stackstate_checks_dev/stackstate_checks/dev/plugin/plugin.py:206]: rename 
`expected_pre_run_state`
`expected_post_run_state`

File: stackstate_checks_base/stackstate_checks/base/checks/base.py [hunk 6]
```diff
@@ -581,9 +649,11 @@ def _get_requests_proxy(self):
 
         return proxies if proxies else no_proxy_settings
 
-    @staticmethod
-    def get_agent_confd_path():
-        return datadog_agent.get_config("confd_path")
+    def get_check_config_path(self):
+        return "{}/{}.d".format(self.get_agent_conf_d_path(), self.name)
```
Comment (TizianoPerrucci) [inline stackstate_checks_base/stackstate_checks/base/checks/base.py:653]: what about windows ?

File: stackstate_checks_base/stackstate_checks/base/utils/persistent_state.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+import json
+import errno
+from schematics import Model
+
+
+class StateDescriptor:
+    """
+    StateDescriptor is used to describe an instance of state that is persisted to disk
+    """
+    def __init__(self, instance_key, check_conf_d_path):
```
Comment (TizianoPerrucci) [inline stackstate_checks_base/stackstate_checks/base/utils/persistent_state.py:15]: do not allow `None` or empty string

**Review by TizianoPerrucci — CHANGES_REQUESTED** [review]

**Review by TizianoPerrucci — APPROVED** [review]

**Pull request merged**